### PR TITLE
System for single device net server per station functionality

### PR DESCRIPTION
**Pull request by juliangiebel**
## About the PR

This PR moves the functionality for one active server entity per station out of the `CrewMonitoringServerSystem` into its own system called `SingletonDeviceNetServerSystem`.

## Why / Balance

This makes it easy to implement other systems that rely on there being one active device net server per station.
Things I plan on implementing in the future will us this.

See:
- [PDA Messaging proposal](https://docs.spacestation14.com/en/proposals/julian-vasilis-pda-messaging.html)
- [DeviceNet ideas](https://hackmd.io/iu2yK9bcQb-veuCOLl-FYw)

## Technical details

The system has a public method for getting the device net address of the currently active server on a station and of a specific type.
The type of server you want is specified using a component type parameter which the server entity needs to contain.

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

### CODE REVIEW

File: Content.Server/DeviceNetwork/Components/SingletonDeviceNetServerComponent.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    [ViewVariables(VVAccess.ReadWrite)]
+    public bool Available = true;
+
+
+    /// <summary>
+    ///     Whether the server is the currently active server for the station it's on
+    /// </summary>
+    [ViewVariables(VVAccess.ReadWrite)]
+    public bool Active = true;
+}
```
Comment (metalgearsloth) [inline Content.Server/DeviceNetwork/Components/SingletonDeviceNetServerComponent.cs:18]: Should be datafields.

File: Content.Server/DeviceNetwork/Systems/SingletonDeviceNetServerSystem.cs [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+/// Server in this context means an entity that manages the devicenet packets like the <see cref="Content.Server.Medical.CrewMonitoring.CrewMonitoringServerSystem"/>
+/// </summary>
+public sealed class SingletonDeviceNetServerSystem : EntitySystem
+{
+    [Dependency] private readonly DeviceNetworkSystem _deviceNetworkSystem = default!;
+    [Dependency] private readonly StationSystem _stationSystem = default!;
+
+    private const float UpdateRate = 3f;
+    private float _updateDiff;
+
```
Comment (metalgearsloth) [inline Content.Server/DeviceNetwork/Systems/SingletonDeviceNetServerSystem.cs:20]: Systems shouldn't store data as it doesn't serialize properly with saves.

File: Content.Server/Medical/CrewMonitoring/CrewMonitoringServerSystem.cs [hunk 3]
```diff
@@ -38,70 +37,15 @@ public override void Update(float frameTime)
...
-            last = (uid, server, device);
-
-            if (server.Active)
-            {
-                address = device.Address;
-                return true;
-            }
-        }
+            if (!_singletonServerSystem.IsActiveServer(id))
+                return;
```
Comment (metalgearsloth) [inline Content.Server/Medical/CrewMonitoring/CrewMonitoringServerSystem.cs:44]:
```suggestion
            if (!_singletonServerSystem.IsActiveServer(id))
                continue;
```

File: Content.Server/DeviceNetwork/Systems/SingletonDeviceNetServerSystem.cs [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        while (servers.MoveNext(out var id, out var server))
+        {
+            //Make sure the server is disconnected when it becomes unavailable
+            if (server.Available)
+                continue;
+
+            if (server.Active)
+                DisconnectServer(id, server);
+        }
+    }
```
Comment (metalgearsloth) [inline Content.Server/DeviceNetwork/Systems/SingletonDeviceNetServerSystem.cs:48]: This looks like it's just hacking around some other state handling by just querying every 3 seconds if a server is active?

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Server/DeviceNetwork/Systems/SingletonDeviceNetServerSystem.cs (hunk 4 above)
Comment (juliangiebel) [reply Content.Server/DeviceNetwork/Systems/SingletonDeviceNetServerSystem.cs]: Yea. I was able to just remove that

File: Content.Server/DeviceNetwork/Systems/SingletonDeviceNetServerSystem.cs (hunk 2 above)
Comment (deltanedas) [reply Content.Server/DeviceNetwork/Systems/SingletonDeviceNetServerSystem.cs]: should probably be stored on a station component (even though thats not serialized :trollface:)

**Pull request merged**